tasks/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+                    // Safety: `unwrap` can't fail here.
+                    let entry = self.waiting_tasks.remove(&handle).unwrap();
+                    self.active_tasks.insert(handle, (tag, entry));
+                }
+                ClientFlowEvent::CommandRejected { tag, status, .. } => {
+                    let handle = {
+                        let (handle, _) = self
+                            .active_tasks
+                            .iter()
+                            .find(|(_, (tag_, _))| tag == *tag_)
+                            .unwrap();
+
+                        *handle
+                    };
+
+                    let body = match StatusV2::from(status) {
+                        StatusV2::Tagged(Tagged { body, .. }) => body,
+                        _ => unreachable!(),
+                    };
+
+                    let (_, task) = self.active_tasks.remove(&handle).unwrap();
+
+                    let output = Some(task.process_tagged(body));
+
+                    return SchedulerEvent::TaskFinished(TaskToken { handle, output });
+                }
+                ClientFlowEvent::DataReceived { data } => {
+                    if let Some(data) = trickle_down(data, self.active_tasks_mut(), |task, data| {
+                        task.process_data(data)
+                    }) {
+                        return SchedulerEvent::Unsolicited(Response::Data(data));
+                    }
+                }
+                ClientFlowEvent::StatusReceived { status } => match StatusV2::from(status) {
+                    StatusV2::Untagged(body) => {
+                        if let Some(body) =
+                            trickle_down(body, self.active_tasks_mut(), |task, body| {
+                                task.process_untagged(body)
+                            })
+                        {
+                            return SchedulerEvent::Unsolicited(Response::Status(Status::from(
+                                StatusV2::Untagged(body),
+                            )));
+                        }
+                    }
+                    StatusV2::Bye(bye) => {
+                        if let Some(bye) =
+                            trickle_down(bye, self.active_tasks_mut(), |task, bye| {
+                                task.process_bye(bye)
+                            })
+                        {
+                            return SchedulerEvent::Unsolicited(Response::Status(Status::from(
+                                StatusV2::Bye(bye),
+                            )));
+                        }
+                    }
+                    StatusV2::Tagged(Tagged { tag, body }) => {
+                        let handle = {
+                            let (handle, _) = self
+                                .active_tasks
+                                .iter()
+                                .find(|(_, (tag_, _))| tag == *tag_)
+                                .unwrap();
```

Comment:
Oh, good catch!

`Unsolicited` would work, but ...

Do we ever give up on a connection?
Like... what needs to happen that we stop a session before we cause damage?

Getting a tagged response to a command we never send is pretty fatal, IMO, as all bets are off.

Reply:
But we don't want to crash the client only because the server is trash. Let the caller decide, we can't give up the connection here. The only thing we can do here is emitting an event with a more seriously sounding name than `Unsolicited`.

Reply:
Oh, you misunderstood. I didn't propose to panic. I never will, in fact :D I was just thinking if we should return, e.g., the `Event::Shutdown` and "fuse" the flow to make it unusable.

Reply:
Let the caller decide :p

But be opinionated by returning a `FatalUnknownTagStatusFromCthulhu` event.